--- a/debian_package/sbin/folder2ram	2016-09-26 16:55:20.608003412 +0200
+++ b/debian_package/sbin/folder2ram	2016-09-26 16:56:48.718899436 +0200
@@ -217,7 +217,7 @@
 
 #writing a tiny amount of stuff to have systemd do the same thing as above.
 #clear symptom that systemd is master race.
-		cat <<'EOF' > "/lib/systemd/system/folder2ram_startup.service"	
+		cat <<'EOF' > "/etc/systemd/system/folder2ram_startup.service"	
 [Unit]
 Description=folder2ram systemd service
 After=local-fs.target
@@ -239,7 +239,7 @@
 
 #writing a tiny amount of stuff to have systemd do the same thing as above.
 #clear symptom that systemd is master race.
-		cat <<'EOF' > "/lib/systemd/system/folder2ram_shutdown.service"	
+		cat <<'EOF' > "/etc/systemd/system/folder2ram_shutdown.service"	
 [Unit]
 Description=folder2ram systemd service
 Before=shutdown.target reboot.target halt.target
@@ -260,7 +260,7 @@
 write_systemd_service_cleanup (){
 
 #asking to delete all the stuff we are leaving behind after we are done with the safe shutdown
-cat <<'EOF' > "/lib/systemd/system/folder2ram_cleaner.service"	
+cat <<'EOF' > "/etc/systemd/system/folder2ram_cleaner.service"	
 [Unit]
 Description=temporary folder2ram systemd service
 After=basic.target
@@ -394,7 +394,7 @@
 	systemd)
 	cat <<'EOF' >> "/sbin/folder2ram_cleaner"
 systemctl disable folder2ram_cleaner.service
-rm -f "/lib/systemd/system/folder2ram_cleaner.service"
+rm -f "/etc/systemd/system/folder2ram_cleaner.service"
 rm -f "/sbin/folder2ram_cleaner.conf"
 rm -f "/sbin/folder2ram_cleaner"
 exit
@@ -1005,13 +1005,13 @@
 		write_systemd_startup_service
 
 		# enabling the service, will be started on reboot
-		systemctl enable "/lib/systemd/system/folder2ram_startup.service"
+		systemctl enable "/etc/systemd/system/folder2ram_startup.service"
 
 		#calling the function that writes the systemd service file
 		write_systemd_shutdown_service
 
 		# enabling the service, will be started on reboot
-		systemctl enable "/lib/systemd/system/folder2ram_shutdown.service"
+		systemctl enable "/etc/systemd/system/folder2ram_shutdown.service"
 
 
 	;;
@@ -1021,14 +1021,14 @@
 		mount_umount_all stop
 
 		#disabling the service
-		systemctl disable "/lib/systemd/system/folder2ram_startup.service"
-		systemctl disable "/lib/systemd/system/folder2ram_shutdown.service"
+		systemctl disable "/etc/systemd/system/folder2ram_startup.service"
+		systemctl disable "/etc/systemd/system/folder2ram_shutdown.service"
 	;;
 	systemd_safe_remove)
 		
 		#disabling the service 
-		systemctl disable "/lib/systemd/system/folder2ram_startup.service"
-		systemctl disable "/lib/systemd/system/folder2ram_shutdown.service"
+		systemctl disable "/etc/systemd/system/folder2ram_startup.service"
+		systemctl disable "/etc/systemd/system/folder2ram_shutdown.service"
 
 		#now we pull the same trick as with the initscript above, but modified to work with systemd
 
