diff -ru gnome-shell-3.22.0.old/data/org.gnome.shell.gschema.xml.in gnome-shell-3.22.0/data/org.gnome.shell.gschema.xml.in
--- gnome-shell-3.22.0.old/data/org.gnome.shell.gschema.xml.in	2016-09-11 20:15:40.000000000 +0300
+++ gnome-shell-3.22.0/data/org.gnome.shell.gschema.xml.in	2016-09-20 16:02:39.999591978 +0300
@@ -31,7 +31,7 @@
       </description>
     </key>
     <key name="favorite-apps" type="as">
-      <default>[ 'epiphany.desktop', 'evolution.desktop', 'rhythmbox.desktop', 'shotwell.desktop', 'org.gnome.Nautilus.desktop', 'org.gnome.Software.desktop' ]</default>
+      <default>[ 'chromium-browser-chromium.desktop', 'evolution.desktop', 'org.gnome.Nautilus.desktop' ]</default>
       <summary>List of desktop file IDs for favorite applications</summary>
       <description>
         The applications corresponding to these identifiers
