--- ./src/network/ssl/qsslcontext_openssl.cpp.orig	2016-09-02 08:57:37.246849675 +0200
+++ ./src/network/ssl/qsslcontext_openssl.cpp	2016-09-02 09:01:40.331907021 +0200
@@ -43,6 +43,12 @@
 #include "private/qsslsocket_openssl_p.h"
 #include "private/qsslsocket_openssl_symbols_p.h"
 
+#if defined(LIBRESSL_VERSION_NUMBER)
+#if !defined(OPENSSL_NO_EC)
+#define OPENSSL_NO_EC
+#endif
+#endif
+
 QT_BEGIN_NAMESPACE
 
 // defined in qsslsocket_openssl.cpp:
