From 31bd644dab64a72bcbfec8bf71225b0264408b56 Mon Sep 17 00:00:00 2001
From: NIIBE Yutaka <gniibe@fsij.org>
Date: Wed, 30 Apr 2014 16:44:19 +0900
Subject: [PATCH 19/52] * MIGRATION, TODO: Apply documentation change from
 Debian.

---
 MIGRATION | 2 ++
 TODO      | 3 +++
 2 files changed, 5 insertions(+)

diff --git a/MIGRATION b/MIGRATION
index 2c33dbc..d5ee697 100644
--- a/MIGRATION
+++ b/MIGRATION
@@ -51,3 +51,5 @@ or greater.
   - "fake-wait-for-card"
   - "require-card-switch"
   - "wait-timeout"
+  - "try_pin"
+  - "quiet"
diff --git a/TODO b/TODO
index c62faea..fe509a1 100644
--- a/TODO
+++ b/TODO
@@ -2,6 +2,9 @@ Final:
 * allow for Dirmngr to be started on demand (in pipe mode) (NO <- Why?!)
 
 Low priority:
+* allow user to skip card authentication without submitting a wrong
+   PIN to the card, e.g. by entering an empty PIN? Return
+   PAM_CRED_INSUFFICIENT in that case? PAM_AUTHINFO_UNAVAIL? PAM_AUTH_ERR?
 * figure out what exactly the dependencies on the OpenPGP smartcard are.
 * improve doc
 * work on MIGRATION text
-- 
2.17.0

