From 61ae5472639e1ee5d751db94a55528dfc640a9d0 Mon Sep 17 00:00:00 2001
From: NIIBE Yutaka <gniibe@fsij.org>
Date: Thu, 1 May 2014 14:08:04 +0900
Subject: [PATCH 29/52] add .gitignore

---
 .gitignore | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)
 create mode 100644 .gitignore

diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..9e1c7bc
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,16 @@
+Makefile.in
+aclocal.m4
+compile
+config.guess
+config.sub
+depcomp
+install-sh
+missing
+autom4te.cache/
+configure
+config.h.in
+doc/poldi.info*
+doc/stamp-vti
+doc/version.texi
+po/*.gmo
+po/stamp-po
\ No newline at end of file
-- 
2.17.0

