From 67d92164f5512d4b0ba5d7461085b7941d091bf8 Mon Sep 17 00:00:00 2001
From: NIIBE Yutaka <gniibe@fsij.org>
Date: Mon, 7 Nov 2016 12:51:16 +0900
Subject: [PATCH 39/52] Add NEWS entries.

--
Signed-off-by: NIIBE Yutaka <gniibe@fsij.org>
---
 NEWS | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/NEWS b/NEWS
index decdc2d..a52f269 100644
--- a/NEWS
+++ b/NEWS
@@ -2,6 +2,19 @@
 
 Changes since version 0.4.1:
 
+* poldi-ctrl is removed
+  Please use gpg-connect-agent instead.
+
+* Poldi always invokes scdaemon to connect it through pipe
+  Older Poldi has a feature of connecting to scdaemon with help of
+  gpg-agent using the GPG_AGENT_INFO enviornment variable.  In GnuPG
+  2.1, the GPG_AGENT_INFO is gone and scdaemon no longer keeps locking
+  the reader after card removal, it is good to always invoke scdaemon
+  for the authentication.  If there is an existing scdaemon with card
+  inserted, a failure is expected and this is safer fallback.  That's
+  because Poldi should not connect to a smartcard which is in use for
+  other purpose and possibly already authenticated.
+
 * New option "scdaemon-options"
   Added a new option "scdaemon-options", which can be used to specify
   the scdaemon configuration file to use for newly spawned scdaemon
-- 
2.17.0

