From e531e32e4179687f4d5ed3081860b4f2c7e628ed Mon Sep 17 00:00:00 2001
From: Luigi 'Comio' Mantellini <luigi.mantellini@gmail.com>
Date: Mon, 11 Jun 2018 17:19:48 +0200
Subject: [PATCH] Add limits.h include

---
 src/scdtotp.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/scdtotp.c b/src/scdtotp.c
index 07d1b22..6a954c2 100644
--- a/src/scdtotp.c
+++ b/src/scdtotp.c
@@ -27,6 +27,7 @@
 #include <time.h>
 #include <getopt.h>
 #include <err.h>
+#include <limits.h>
 
 #include <gcrypt.h>
 
-- 
2.17.1

