diff --git a/Makefile b/Makefile
index 0b3837c..537aed0 100644
--- a/Makefile
+++ b/Makefile
@@ -83,7 +83,8 @@ CXXFLAGS    = $(OPTS) \
               -DFUSE_USE_VERSION=29 \
               -MMD \
 	      -DUSE_XATTR=$(USE_XATTR) \
-	      -DUGID_USE_RWLOCK=$(UGID_USE_RWLOCK)
+	      -DUGID_USE_RWLOCK=$(UGID_USE_RWLOCK) \
+	      $(CXXFLAGS)
 
 PREFIX        = /usr/local
 EXEC_PREFIX   = $(PREFIX)
