diff --git a/Makefile b/Makefile
index 9d54d61..3f6d949 100644
--- a/Makefile
+++ b/Makefile
@@ -60,13 +60,11 @@ endif
 
 UGID_USE_RWLOCK = 0
 
-OPTS 	    = -O2
 SRC	    = $(wildcard src/*.cpp)
 OBJS        = $(SRC:src/%.cpp=build/%.o)
 DEPS        = $(SRC:src/%.cpp=build/%.d)
 MANPAGE     = mergerfs.1
 CXXFLAGS   += \
-	      $(OPTS) \
               $(DEBUG_FLAGS) \
               $(STATIC_FLAGS) \
               $(LTO_FLAGS) \
